### DISCUSSION TITLE
Large memory consumption when changing client request body

### DISCUSSION THREAD

**Discussion by ityuhui** [nginxinc/nginx-s3-gateway#140] [Q&A]
Hello !

Thank you for creating and maintainting this great project !
I'm working to support chunked upload based on this project. Because the size of my client request body is large (512MB or bigger), I enabled the temp files:

```
    client_max_body_size            0;
    client_body_temp_path           /tmp/client_temp;
    client_body_in_file_only        clean;
```

And I need to resign the chunk-signatures for chunked data, so I set "proxy_set_body"

```
    js_set $s3body s3gateway.s3body;
    proxy_set_body $s3body;
```

Here's the explanation of my s3 body functions:
```Javascript
function s3body(r) {
    if (typeof r.variables.request_body_file == 'undefined') {
        return "";
    }

    /**
    This is a chunk body that needs re-sign with the real crends. e.g.
    ```
    2a;chunk-signature=af811297ad3747879541b351afc7f67531ae810cc0f671a43bcd50de9807f4ff
    firstname,lastname,goals
    12345,1234567,22

    0;chunk-signature=5e6d2b065095aed06a9a2c0707145d516036211011ba3f8e7273ec264acf46ad

    ```
    */

    Read a chunked data from the temp file, calculate and replace the chunk signature, append to a string (resignedBody) repeatly. 

    return resignedBody;
}
```

Everything works fine.
But the memory consumpation is very large. In my test, I uploaded 17 MB data, the memory consumption of the nginx-s3-gateway docker container was 250 MB. 
Even though the upload is completed, the memory is not freed.
When the next upload starts, nginx reuses this memory, so the memory consumption does not keep increasing. So I'm guessing this is in a memory pool allocated by Nginx for NJS and not returned to the system after a request is completed.

My question is:
What is wrong with my implementation? How to reduce memory consumption?
Thank you!

**Comment by dekobon** [comment]
For a comparison, what does memory consumption look like on a similarly configured instance that is not running the s3 gateway njs code? As I understand it, NGINX memory utilization is quite variable and the OS is not quick to reclaim memory that is reclaimable. There are a number of things going on that may or may not be issues.

Also, NJS has no garbage collection. Memory is reclaimed only after a request finishes. The OS may not actually reclaim this memory until it decides to do so, leaving more memory allocated to NGINX.

**Comment by ityuhui** [comment]
Thank you for your reply.
